oak-run-commons/src/test/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/MergeRunnerTest.java

```diff
@@ -0,0 +1,176 @@
+                add(testFiles.get(6));
+            }
+        }));
+        expectedLogOutput.add("created merge task for intermediate-2 with " + (new ArrayList<File>(){
+            {
+                add(testFiles.get(3));
+                add(testFiles.get(4));
+                add(testFiles.get(7));
+            }
+        }));
+        expectedLogOutput.add("created merge task for intermediate-3 with " + (new ArrayList<File>(){
+            {
+                add(testFiles.get(1));
+                add(testFiles.get(2));
+                add(testFiles.get(8));
+            }
+        }));
+        expectedLogOutput.add("Waiting for batch sorting tasks completion");
+        assertEquals("intermediate merge log output should be expected", String.join(newline, expectedLogOutput), String.join(newline, actualLogOutput.subList(0,4)));
+
+        assertTrue("merge complete output should exist", actualLogOutput.containsAll((new ArrayList<String>(){
+            {
+                add("merge complete for intermediate-1");
+                add("merge complete for intermediate-2");
+                add("merge complete for intermediate-3");
+            }
+        })));
+
+        expectedLogOutput.clear();
+        expectedLogOutput.add("There are still 4 sorted files not merged yet");
+        expectedLogOutput.add("running final batch merge task for final-1 with " + (new ArrayList<File>(){
+            {
+                add(testFiles.get(9));
+                add(testFiles.get(10));
+                add(testFiles.get(11));
+            }
+        }));
+        expectedLogOutput.add("running final batch merge task for final-2 with " + (new ArrayList<File>(){
+            {
+                add(testFiles.get(12));
+                add(new File(mergeDir, "intermediate-1"));
+                add(new File(mergeDir, "intermediate-3"));
+            }
+        }));
+        expectedLogOutput.add("running final batch merge task for " + sortedFile.getName() + " with " + (new ArrayList<File>(){
+            {
+                add(new File(mergeDir, "intermediate-2"));
+                add(new File(mergeDir, "final-1"));
+                add(new File(mergeDir, "final-2"));
+            }
+        }));
+        expectedLogOutput.add("Total batch sorted files length is 18");
+        assertEquals("final merge log output should be expected", String.join(newline, expectedLogOutput), String.join(newline, actualLogOutput.subList(7, 12)));
+    }
+
+    private void generateTestFiles(File tmpDir) throws IOException {
+        int nextFileSize = 1;
+        File testFile = new File(tmpDir, Integer.toString(nextFileSize));
+        List<String> lineBuffer = Lists.newArrayList();
+        List<String> resultList = Lists.newArrayList();
+        for (int i = 0; i <= 90; i++) {
+            String line = String.format("/%05d|{}", i);
+            lineBuffer.add(line);
+            resultList.add(line);
+            if (lineBuffer.size() == nextFileSize) {
```

Comment:
You loop 90 times, and if the buffer size matches the next file size, you switch to the new file... but it's not easy (for me) to understand how many files are generated that way... Wouldn't it be easier to understand to loop over the list of files?

What about making the files smaller and smaller? That way, it would be easy to verify they are merged in the right order (just an idea).

Reply:
did something similar but not exactly the same.
change to use nodeState for generating the lines.
so if fileCount is 3 => we get file with name `1`, `2`, `3`
and each file `1` contains 1 line, file `2` contains 2 lines and so on.
line contents are randomized but sorted.